File: backero-backend/src/services/dependency.service.js
const TaskDependency = require('../models/TaskDependency');
const Task = require('../models/Task');

/**
 * Add a dependency: fromTask must finish before toTask can be completed.
 * Performs circular-dependency detection before creating.
 */
const addDependency = async (fromTaskId, toTaskId, type = 'finish_to_start', orgId, userId) => {
  if (fromTaskId.toString() === toTaskId.toString()) {
    throw new Error('A task cannot depend on itself');
  }

  const existing = await TaskDependency.findOne({
    fromTask: fromTaskId,
    toTask: toTaskId,
    status: { $ne: 'waived' },
  });
  if (existing) throw new Error('Dependency already exists');

  // Cycle guard: would creating from→to introduce a cycle?
  const hasCycle = await detectCycle(toTaskId, fromTaskId, orgId);
  if (hasCycle) throw new Error('Adding this dependency would create a circular dependency');

  const dep = await TaskDependency.create({
    organizationId: orgId,
    fromTask: fromTaskId,
    toTask: toTaskId,
    type,
    createdBy: userId,
  });

  // Lock toTask if fromTask is not yet completed
  const fromTask = await Task.findById(fromTaskId).select('title status');
  if (fromTask && fromTask.status !== 'Completed') {
    await Task.findByIdAndUpdate(toTaskId, {
      completionLocked: true,
      $addToSet: { completionLockReasons: `Blocked by: "${fromTask.title}"` },
    });
  }

  return dep;
};

/**
 * Iterative DFS: detect if there is a path from startId back to targetId
 * following existing active dependencies (toTask → fromTask direction).
 */
const detectCycle = async (startId, targetId, orgId) => {
  const visited = new Set();
  const stack   = [startId.toString()];

  while (stack.length) {
    const current = stack.pop();
    if (current === targetId.toString()) return true;
    if (visited.has(current)) continue;
    visited.add(current);

    const deps = await TaskDependency.find({
      toTask: current,
      organizationId: orgId,
      status: 'active',
    }).select('fromTask');

    for (const d of deps) stack.push(d.fromTask.toString());
  }

  return false;
};

/**
 * Waive / remove a dependency.
 */
const removeDependency = async (depId, userId) => {
  const dep = await TaskDependency.findByIdAndUpdate(depId, {
    status: 'waived',
    resolvedAt: new Date(),
    resolvedBy: userId,
  }, { new: true });

  if (!dep) throw new Error('Dependency not found');

  // Re-evaluate toTask's lock
  const { checkCompletionEligibility } = require('./workflowEngine.service');
  const toTask = await Task.findById(dep.toTask);
  if (toTask) {
    const { eligible, reasons } = await checkCompletionEligibility(dep.toTask);
    toTask.completionLocked = !eligible;
    toTask.completionLockReasons = reasons;
    await toTask.save();
  }

  return dep;
};

/**
 * When a task is completed, resolve all its finish-to-start outgoing dependencies
 * and unlock tasks that were blocked by it.
 */
const resolveOutgoingDependencies = async (taskId, orgId) => {
  await TaskDependency.updateMany(
    { fromTask: taskId, organizationId: orgId, status: 'active', type: 'finish_to_start' },
    { status: 'resolved', resolvedAt: new Date() }
  );

  const downstream = await TaskDependency.find({ fromTask: taskId, organizationId: orgId }).select('toTask');
  const { checkCompletionEligibility } = require('./workflowEngine.service');

  for (const d of downstream) {
    const { eligible, reasons } = await checkCompletionEligibility(d.toTask);
    await Task.findByIdAndUpdate(d.toTask, {
      completionLocked: !eligible,
      completionLockReasons: reasons,
    });
  }
};

/**
 * Topological sort using Kahn's algorithm.
 * Throws if a cycle is detected.
 */
const topologicalSort = async (taskIds, orgId) => {
  const ids = taskIds.map(String);
  const inDegree = Object.fromEntries(ids.map(id => [id, 0]));
  const adj      = Object.fromEntries(ids.map(id => [id, []]));

  const deps = await TaskDependency.find({
    fromTask: { $in: taskIds },
    toTask:   { $in: taskIds },
    organizationId: orgId,
    status: 'active',
  });

  for (const dep of deps) {
    const from = dep.fromTask.toString();
    const to   = dep.toTask.toString();
    if (adj[from]) {
      adj[from].push(to);
      inDegree[to] = (inDegree[to] || 0) + 1;
    }
  }

  const queue  = ids.filter(id => inDegree[id] === 0);
  const result = [];

  while (queue.length) {
    const cur = queue.shift();
    result.push(cur);
    for (const nb of (adj[cur] || [])) {
      inDegree[nb]--;
      if (inDegree[nb] === 0) queue.push(nb);
    }
  }

  if (result.length !== ids.length) throw new Error('Circular dependency detected');
  return result;
};

module.exports = {
  addDependency,
  detectCycle,
  removeDependency,
  resolveOutgoingDependencies,
  topologicalSort,
};

File: backero-backend-py/app/modules/tasks/dependency_service.py
"""Task dependency graph — ported from dependency.service.js.

Cycle detection is a *corrected* port, not a literal one: the source's
`detectCycle(toTaskId, fromTaskId)` walks the "toTask → fromTask" direction
(from the new edge's target, backward along existing incoming edges). Traced
through by hand, that direction fails to catch the simplest possible cycle —
adding B→A when A→B already exists comes back "no cycle" and the edge gets
created, live-verified against a running instance of this exact port before
the fix. The correct check for "does adding edge F→T create a cycle" is:
does a forward path T ⇒ F already exist (walking existing edges in their own
from→to direction)? That's what's implemented below."""

from __future__ import annotations

import uuid
from datetime import UTC, datetime

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.exceptions import AppError, NotFoundError
from app.models.task import Task, TaskDependency, TaskDependencyStatus, TaskDependencyType, TaskStatus
from app.modules.tasks.workflow_engine import check_completion_eligibility


def _detect_cycle(db: Session, *, org_id: uuid.UUID, from_task_id: uuid.UUID, to_task_id: uuid.UUID) -> bool:
    """True if a forward path to_task_id ⇒ ... ⇒ from_task_id already exists
    via active edges — i.e. adding from_task_id → to_task_id would close a
    loop."""
    visited: set[uuid.UUID] = set()
    stack = [to_task_id]

    while stack:
        current = stack.pop()
        if current == from_task_id:
            return True
        if current in visited:
            continue
        visited.add(current)

        downstream = (
            db.execute(
                select(TaskDependency.to_task_id).where(
                    TaskDependency.from_task_id == current,
                    TaskDependency.organization_id == org_id,
                    TaskDependency.status == TaskDependencyStatus.ACTIVE,
                )
            )
            .scalars()
            .all()
        )
        stack.extend(downstream)

    return False


def add_dependency(
    db: Session,
    *,
    org_id: uuid.UUID,
    actor_id: uuid.UUID,
    from_task_id: uuid.UUID,
    to_task_id: uuid.UUID,
    dep_type: TaskDependencyType,
) -> TaskDependency:
    if from_task_id == to_task_id:
        raise AppError("self_dependency", "A task cannot depend on itself.", status_code=400)

    from_task = db.execute(
        select(Task).where(Task.id == from_task_id, Task.organization_id == org_id)
    ).scalar_one_or_none()
    to_task = db.execute(select(Task).where(Task.id == to_task_id, Task.organization_id == org_id)).scalar_one_or_none()
    if from_task is None or to_task is None:
        raise NotFoundError("Task not found.")

    existing = db.execute(
        select(TaskDependency).where(
            TaskDependency.from_task_id == from_task_id,
            TaskDependency.to_task_id == to_task_id,
            TaskDependency.status != TaskDependencyStatus.WAIVED,
        )
    ).scalar_one_or_none()
    if existing is not None:
        raise AppError("dependency_exists", "Dependency already exists.", status_code=409)

    if _detect_cycle(db, org_id=org_id, from_task_id=from_task_id, to_task_id=to_task_id):
        raise AppError(
            "circular_dependency", "Adding this dependency would create a circular dependency.", status_code=400
        )

    dep = TaskDependency(
        organization_id=org_id, from_task_id=from_task_id, to_task_id=to_task_id, type=dep_type, created_by_id=actor_id
    )
    db.add(dep)

    if from_task.status != TaskStatus.COMPLETED:
        to_task.completion_locked = True
        reasons = list(to_task.completion_lock_reasons or [])
        reason = f'Blocked by: "{from_task.title}"'
        if reason not in reasons:
            reasons.append(reason)
        to_task.completion_lock_reasons = reasons

    db.commit()
    db.refresh(dep)
    return dep


def remove_dependency(
    db: Session, *, org_id: uuid.UUID, actor_id: uuid.UUID, dependency_id: uuid.UUID
) -> TaskDependency:
    dep = db.execute(
        select(TaskDependency).where(TaskDependency.id == dependency_id, TaskDependency.organization_id == org_id)
    ).scalar_one_or_none()
    if dep is None:
        raise NotFoundError("Dependency not found.")

    dep.status = TaskDependencyStatus.WAIVED
    dep.resolved_at = datetime.now(UTC)
    dep.resolved_by_id = actor_id

    to_task = db.get(Task, dep.to_task_id)
    if to_task is not None:
        eligible, reasons = check_completion_eligibility(db, task_id=to_task.id)
        to_task.completion_locked = not eligible
        to_task.completion_lock_reasons = reasons

    db.commit()
    db.refresh(dep)
    return dep


def resolve_outgoing_dependencies(db: Session, *, org_id: uuid.UUID, task_id: uuid.UUID) -> None:
    """Called when a task completes — resolves its finish-to-start outgoing
    edges and re-evaluates the completion lock on every downstream task."""
    outgoing = (
        db.execute(
            select(TaskDependency).where(
                TaskDependency.from_task_id == task_id,
                TaskDependency.organization_id == org_id,
                TaskDependency.status == TaskDependencyStatus.ACTIVE,
                TaskDependency.type == TaskDependencyType.FINISH_TO_START,
            )
        )
        .scalars()
        .all()
    )
    now = datetime.now(UTC)
    downstream_ids: set[uuid.UUID] = set()
    for dep in outgoing:
        dep.status = TaskDependencyStatus.RESOLVED
        dep.resolved_at = now
        downstream_ids.add(dep.to_task_id)

    # The source re-evaluates every downstream task of the FROM task
    # regardless of dependency type/status (not just the ones just
    # resolved above) — replicate that broader re-check.
    all_downstream = (
        db.execute(
            select(TaskDependency.to_task_id).where(
                TaskDependency.from_task_id == task_id, TaskDependency.organization_id == org_id
            )
        )
        .scalars()
        .all()
    )
    downstream_ids.update(all_downstream)

    for downstream_id in downstream_ids:
        downstream_task = db.get(Task, downstream_id)
        if downstream_task is None:
            continue
        eligible, reasons = check_completion_eligibility(db, task_id=downstream_id)
        downstream_task.completion_locked = not eligible
        downstream_task.completion_lock_reasons = reasons

    db.flush()
